### PR TITLE
XIVY-11890: Add known Custom Fields to Table

### PR DESCRIPTION
**Pull request by ivy-edp**
I added Combocells to the Custom Fields Table, inwhich known Custom Fields can be selected. If a known Custom Field is already in the Table it is no longer displayed in the dropdown. Also after selecting a Custom Field via the dropdown the type is also automatically adjusted.
There are UT I had to skip, because the test didn't work anymore with the combobox. Will fix it in a later PR.
beispiel_customfields

### CODE REVIEW

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 1]
```diff
@@ -32,12 +34,43 @@ const EMPTY_WFCUSTOMFIELD: WfCustomField = { name: '', type: 'STRING', value: ''
...
+      return 'TEXT';
+    }
+    if (typeAsString === 'NUMBER') {
+      return 'NUMBER';
+    }
+    if (typeAsString === 'TIMESTAMP') {
+      return 'TIMESTAMP';
+    }
+    return 'STRING';
+  };
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:51]: looks like a useless conversion to me... Why is this needed?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 2]
```diff
@@ -32,12 +34,43 @@ const EMPTY_WFCUSTOMFIELD: WfCustomField = { name: '', type: 'STRING', value: ''
...
+
+  const predefinedCustomField: WfCustomField[] = useMeta(
+    'meta/workflow/customFields',
+    { context, type: toWorkflowType(path) },
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:55]: why? I think the path never be only 'Text' or 'Number' or 'Timestamp', so it always results in String...

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 3]
```diff
@@ -32,12 +34,43 @@ const EMPTY_WFCUSTOMFIELD: WfCustomField = { name: '', type: 'STRING', value: ''
...
+    []
+  ).data.map<WfCustomField>(pcf => ({
+    name: pcf.name,
+    type: toWfFieldType(pcf.type),
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:59]: what does pcf mean?
why do we have to convert this type?

File: packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx [hunk 4]
```diff
@@ -17,20 +28,30 @@ type StartCustomFieldTableProps = {
 const EMPTY_STARTCUSTOMSTARTFIELD: StartCustomStartField = { name: '', value: '' } as const;
 
 const StartCustomFieldTable = ({ data, onChange }: StartCustomFieldTableProps) => {
+  const { context } = useEditorContext();
+  const path = usePath();
+  const predefinedCustomField: ComboboxItem[] = useMeta(
+    'meta/workflow/customFields',
+    { context, type: toWorkflowType(path) },
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx:35]: it is always type: 'START' here right?

File: packages/editor/src/components/parts/common/table/useResizableEditableTable.ts [hunk 5]
```diff
@@ -31,20 +38,24 @@ const useResizableEditableTable = <TData>({ data, columns, onChange, emptyDataOb
...
+                [columnId]: value
+              };
+            }
+            return row;
+          });
+          if (!deepEqual(updatedData[updatedData.length - 1], emptyDataObject) && rowIndex === data.length - 1) {
+            onChange([...updatedData, emptyDataObject]);
+          } else {
+            onChange(updatedData);
+          }
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/table/useResizableEditableTable.ts:58]: instead of blow this up, I would create a updateData function outside of this meta object and override it with the specialUpdateData stuff if available... Or maybe even more can be shared between those two?

File: packages/protocol/src/data/inscription.ts [hunk 6]
```diff
@@ -209,6 +206,11 @@ export interface ContentObject {
 export interface MapStringString {
   [k: string]: string;
 }
+export interface CustomFieldMeta {
+  description: string;
+  name: string;
+  type: string;
```
Comment (ivy-lli) [inline packages/protocol/src/data/inscription.ts:212]: this should maybe more typed?

Comment (ivy-lli) [reply packages/protocol/src/data/inscription.ts:212]: I think we know what different types here exist right?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx (hunk 2 above)
Comment (ivy-lli) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:55]: and we already have the type of the customfield table in the props right?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx (hunk 1 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx]: Yes, I have now removed that conversion. The version you reviewed still has an outdated inscription.ts. Now I receive a WfCustomField array directly from "meta/workflow/customFields". Accordingly, I no longer need this. Sorry, I meant to add this as a reviewer for this PR just now.

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx (hunk 2 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx]: I am not sure if I understand correctly, but in this path, it's not about Text, Number, or Timestamp, but rather 'START, CASE, TASK'. This is necessary because there are different predefined custom fields depending on whether you're on a start, case, or task.

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx (hunk 3 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx]: No longer needed. I removed this mapping 👍

File: packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx (hunk 4 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx:35]: Yes, you're right, the StartCustomFieldTable is effectively only used in the RequestPart. Should I just always set START as the type in this case?

File: packages/protocol/src/data/inscription.ts (hunk 6 above)
Comment (ivy-edp) [reply packages/protocol/src/data/inscription.ts]: Does not exist anymore 👍. Instead of CustomFieldMeta, there is now customFieldSchema, which is of type "WfCustomField". If you're interested, this is the PR for the core changes.
https://github.com/axonivy/core/pull/5006

File: packages/editor/src/components/parts/common/table/useResizableEditableTable.ts (hunk 5 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/table/useResizableEditableTable.ts:58]: do you mean something like this?
grafik

Comment (ivy-lli) [reply packages/editor/src/components/parts/common/table/useResizableEditableTable.ts]: Yes I had something like this in mind

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 7]
```diff
@@ -1,43 +1,55 @@
...
 
 const EMPTY_WFCUSTOMFIELD: WfCustomField = { name: '', type: 'STRING', value: '' } as const;
 
 const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
   const items = useMemo<SelectItem[]>(() => Object.entries(CUSTOM_FIELD_TYPE).map(([value, label]) => ({ label, value })), []);
 
+  const { context } = useEditorContext();
+  const path = usePath();
+
+  const predefinedCustomField: WfCustomField[] = useMeta('meta/workflow/customFields', { context, type: toWorkflowType(path) }, []).data;
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:40]: toWorkflowType IMO not needed, type is already there from the component props

File: packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx (hunk 4 above)
Comment (ivy-lli) [reply packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx:35]: yes

File: packages/editor/src/components/parts/common/info/Information.tsx
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/info/Information.tsx:1]: changes not needed, as you don't need to export the `toWorkflowType` I think

**Review by ivy-lli — COMMENTED** [review]
Much better, but I still think you can simply use the prop type and don't need this toWorkflowType method...

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx (hunk 7 above)
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:40]: Ah now I get it, thanks. I will remove the toWorkflowType method from CustomFieldTable 👍

**Review by ivy-lli — APPROVED** [review]
